odoo/addons/base/models/res_users.py

```diff
@@ -199,6 +199,61 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
             return len(groups) if count else groups.ids
         return super(Groups, self)._search(args, offset=offset, limit=limit, order=order, count=count, access_rights_uid=access_rights_uid)
 
+    def _get_implied_groups(self):
+        current_groups = self
+        implied_groups = self.env['res.groups']
+        while current_groups:
+            new_implied_groups = current_groups.implied_ids.filtered(
+                lambda grp: grp.category_id == self.category_id and grp not in implied_groups
+            )
+            implied_groups |= new_implied_groups
+            current_groups = new_implied_groups
+        return implied_groups
+
+    def _get_warning_message(self, user, group, category, possible_groups):
+        return _(
+            "Since %s is a/an %s %s, you can set %s right only to %s",
+            user.name,
+            group.category_id.name,
+            group.name,
+            category,
+            _(" or ").join(possible_groups.mapped('name'))
+        )
+
+    def _prepare_warning_for_group_inheritance(self, user):
+        # Current groups of the user
+        current_groups = user.groups_id._origin
+        # Groups that user had
+        all_groups = user._origin.groups_id
+        # All implied groups
+        implied_groups = (current_groups.implied_ids | all_groups.implied_ids) - current_groups
+        warnings = []
+        category_read_groups = self._read_group(
+            domain=[('category_id', 'in', all_groups.category_id.ids)],
+            fields=['ids:array_agg(id)'],
+            groupby=['category_id'],
+        )
+        for group in current_groups:
+            # Get the minimum group needed in selected group's category (For example, someone changes
+            # Sales: Admin to Sales: User, but Field Service is already set to Admin, so here in the
+            # 'Sales' category, we will at the minimum need Admin group)
+            minimum_required_groups = implied_groups.filtered(
+                lambda grp:
+                grp in group.implied_ids and
+                grp.category_id != group.category_id and
+                grp not in current_groups.filtered(
+                    lambda group:
+                    group.category_id == grp.category_id).implied_ids
+            )
+            for minimum_required_group in minimum_required_groups:
+                category_groups_list = list(filter(lambda group: group['category_id'][0] == minimum_required_group.category_id.id, category_read_groups))
+                category_groups = self.browse(category_groups_list[0]['ids'])
+                lower_groups = minimum_required_group._get_implied_groups()
+                # Groups that have higher access level than minimum_required_groups
+                possible_groups = minimum_required_group | (category_groups - lower_groups)
+                warnings.append(self._get_warning_message(user, group, minimum_required_group.category_id.name, possible_groups))
+        return warnings
```

Comment:
I think we can simplify even more, if we accept to have only the first group (so no `OR blabla` in the warning message)

```suggestion
    def _prepare_warning_for_group_inheritance(self, user):
        # Current groups of the user
        current_groups = user.groups_id

        current_groups_by_category = defaultdict(lambda: self.env['res.groups'])
        for group in current_groups:
            current_groups_by_category[group.category_id] |= group

        missing_groups = []
        for group in current_groups:
            missing_implied_groups = group.implied_ids - current_groups
            missing_implied_groups = missing_implied_groups.filtered(
                lambda g:
                    g.category_id != group.category_id
                    and g not in current_groups_by_category[g.category_id].implied_ids
            )

            # add missing group, by categories
            for missing_implied_group in missing_implied_groups:
                missing_groups.append((missing_implied_group, group))

        return [
            self._get_warning_message(user, group, group_required.category_id.name, group_required)
            for group_required, group in missing_groups
        ]
```